### PR TITLE
Add support for many missing raw formats

### PR DESCRIPTION
**Pull request by elliotclee**
Now that libraw support is in (thanks @uhthomas), as per issue #2745 this PR adds uploading of most of the raw photo file formats supported by libraw. While not every last possible format was added (libraw does not have a definitive list of what it supports), I got all the ones mentioned by gnome-raw-thumbnailer and most of the ones mentioned by digikam. Also allows uploading of .avif files (#2803 ).

Feedback welcome!

### CODE REVIEW

**Comment by uhthomas** [comment]
Looks awesome, thanks @elliotclee! Have you had a chance to test any of these formats?

File: server/src/immich/config/profile-image-upload.config.ts [hunk 1]
```diff
@@ -25,7 +25,7 @@ function fileFilter(req: AuthRequest, file: any, cb: any) {
     return cb(new UnauthorizedException());
   }
 
-  if (file.mimetype.match(/\/(jpg|jpeg|png|heic|heif|dng|webp)$/)) {
+  if (file.mimetype.match(/\/(jpg|jpeg|png|heic|heif|dng|webp|avif)$/)) {
```
Comment (uhthomas) [inline server/src/immich/config/profile-image-upload.config.ts:28]: So this check is for profile images. I don't know what restrictions there are or why. Would be good to check.

File: web/src/lib/utils/asset-utils.ts [hunk 2]
```diff
@@ -131,11 +131,35 @@ export function getFileMimeType(file: File): string {
...
+		cr2: 'image/x-canon-cr2',
+		cr3: 'image/x-canon-cr3',
+		erf: 'image/x-epson-erf',
+		raf: 'image/x-fuji-raf',
+		dcr: 'image/x-kodak-dcr',
+		k25: 'image/x-kodak-k25',
+		kdc: 'image/x-kodak-kdc',
+		mrw: 'image/x-minolta-mrw',
+		orf: 'image/x-olympus-orf',
+		raw: 'image/x-panasonic-raw',
```
Comment (klejejs) [inline web/src/lib/utils/asset-utils.ts:150]: I have some `rw2` files that, as far as I understand, are the same `x-panasonic-raw` files. Could this extension be added as well?

Comment (elliotclee) [reply web/src/lib/utils/asset-utils.ts:150]: I looked into this. .rw2 files are a bit different than .raw files, and libraw does not presently support them according to https://www.libraw.org/node/2710 . The format wasn't on the digikam or gnome-raw-thumbnailer list of formats supported by libraw.

File: server/src/immich/config/profile-image-upload.config.ts (hunk 1 above)
Comment (elliotclee) [reply server/src/immich/config/profile-image-upload.config.ts:28]: Yea, I wasn't sure either, but I figured "if HEIF is there, and DNG is there, AVIF can be there too" :-)

I saw @alextran1502 mention somewhere that AVIF couldn't be used across the board for things like thumbnail images due to inconsistent AVIF support across the various platforms. Does anyone know more details about this so we have a better idea of what limitations we have to cope with?

**Comment by alextran1502** [comment]
Hello @elliotclee, thank you for the contribution. For each of the new file types added. Have you had a chance to test them to see if thumbnails can be generated from those files?

**Comment by elliotclee** [comment]
@uhthomas and @alextran1502 both asked about testing all these formats. In many cases I just don't have these file formats handy. My collection of raw images is limited to .cr2. Do we need to create a library of images in different formats, and add it to a test or something in order to exercise libraw etc.? Or is that beyond the scope of the PR?

**Comment by alextran1502** [comment]
So I think we can leave the work here since they are supported by `libraw`, if any issue comes up with a specific file format that the thumbnail cannot be generated, we can look at removing them.

**Comment by alextran1502** [comment]
Can you help fix the format of the server and web?

**Comment by elliotclee** [comment]
Yes, I will work on getting the tests to pass when I have a chance to work on this next.

**Review by alextran1502 — APPROVED** [review]

**Comment by Thoroslives** [comment]
RAF format is not supporte, shown as broken images. In thumbnail and in full screen view
image

**Comment by AChep** [comment]
.cr2 also shows up as broken images. Here's a sample file:
[IMG_8302.zip](https://github.com/immich-app/immich/files/11956647/IMG_8302.zip)

**Comment by uhthomas** [comment]
@AChep Yeah :/ This is because of https://github.com/libvips/libvips/pull/3460

**Comment by herobrauni** [comment]
Olympus .orf is also broken in thumbnail and in full screen view
I can provide sample file(s) if needed.

**Comment by shant1010** [comment]
Same, I'm having issues with CR2 files as well.

**Comment by trampe** [comment]
I am also having the exact same issue with CR2 files. I have tens of thousands of Canon .CR2 RAW images that I'd love to upload into Immich.
I'm not a TS, Svelte, or Dart developer, but is there something I can do to help troubleshoot this issue?

**Comment by uhthomas** [comment]
> I am also having the exact same issue with CR2 files. I have tens of thousands of Canon .CR2 RAW images that I'd love to upload into Immich.
> 
> I'm not a TS, Svelte, or Dart developer, but is there something I can do to help troubleshoot this issue?

This is blocked on libvips and imagemagick. I think some work has been done to support CR2 and CR2 files, so we may just need to update the packages.

**Comment by tebra-jl** [comment]
Same for me, .cr2 shows up as broken images with latest version of immich (dockers on debian arm64).

**Comment by tebra-jl** [comment]
> > I am also having the exact same issue with CR2 files. I have tens of thousands of Canon .CR2 RAW images that I'd love to upload into Immich.
> > I'm not a TS, Svelte, or Dart developer, but is there something I can do to help troubleshoot this issue?
> 
> This is blocked on libvips and imagemagick. I think some work has been done to support CR2 and CR2 files, so we may just need to update the packages.

Hi @uhthomas , did you know if the latest version of immich (v1.72.1) have updated packages for CR2 thumbnails (libvips and imagemagick) ?

**Comment by carpikes** [comment]
> > > I am also having the exact same issue with CR2 files. I have tens of thousands of Canon .CR2 RAW images that I'd love to upload into Immich.
> > > I'm not a TS, Svelte, or Dart developer, but is there something I can do to help troubleshoot this issue?
> > 
> > 
> > This is blocked on libvips and imagemagick. I think some work has been done to support CR2 and CR2 files, so we may just need to update the packages.
> 
> Hi @uhthomas , did you know if the latest version of immich (v1.72.1) have updated packages for CR2 thumbnails (libvips and imagemagick) ?

I just installed immich v1.72.2 and the CR2 support is still broken :(

**Comment by trampe** [comment]
> I just installed immich v1.72.2 and the CR2 support is still broken :(
Same thing here. I apologize for my naivety. Does blocked mean that we're waiting for libvips and/or imagemagick to be updated? Why did it work before?

Thanks!

**Comment by coissac** [comment]
Hi,

Same problem for me with version 1.72.2. If it helps, I've tested from the immich_microserver container console, the 'magick' command to convert .CR2 files to jpeg and webp works (except color management, which isn't perfect, but that's a detail).

Thank you

Eric

**Comment by coissac** [comment]
Another test using the `vips` command-line tool. This time the conversion doesn't work, as `vips` still seems to be confusing *CR2* and *TIFF* formats. See the log below (still from the same microserver console)

```bash
vips copy IMG_5017.CR2 IMG_5017.jpeg

(vips:84): VIPS-WARNING **: 06:12:34.078: Photometric tag is missing, assuming data is YCbCr

(vips:84): VIPS-WARNING **: 06:12:34.078: SamplesPerPixel tag is missing, applying correct SamplesPerPixel value of 3

(vips:84): VIPS-WARNING **: 06:12:34.078: Subsampling tag is not set, yet subsampling inside JPEG data [2,1] does not match default values [2,2]; assuming subsampling inside JPEG data is correct

(vips:84): VIPS-WARNING **: 06:12:34.078: Photometric tag is missing, assuming data is YCbCr

(vips:84): VIPS-WARNING **: 06:12:34.078: SamplesPerPixel tag is missing, applying correct SamplesPerPixel value of 3

(vips:84): VIPS-WARNING **: 06:12:34.078: Subsampling tag is not set, yet subsampling inside JPEG data [2,1] does not match default values [2,2]; assuming subsampling inside JPEG data is correct

(vips:84): VIPS-WARNING **: 06:12:34.078: Photometric tag is missing, assuming data is YCbCr

(vips:84): VIPS-WARNING **: 06:12:34.078: SamplesPerPixel tag is missing, applying correct SamplesPerPixel value of 3

(vips:84): VIPS-WARNING **: 06:12:34.078: Subsampling tag is not set, yet subsampling inside JPEG data [2,1] does not match default values [2,2]; assuming subsampling inside JPEG data is correct

(vips:84): VIPS-WARNING **: 06:12:34.078: Photometric tag is missing, assuming data is YCbCr

(vips:84): VIPS-WARNING **: 06:12:34.078: SamplesPerPixel tag is missing, applying correct SamplesPerPixel value of 3

(vips:84): VIPS-WARNING **: 06:12:34.078: Subsampling tag is not set, yet subsampling inside JPEG data [2,1] does not match default values [2,2]; assuming subsampling inside JPEG data is correct

(vips:84): VIPS-WARNING **: 06:12:34.079: Unknown field with tag 513 (0x201) encountered

(vips:84): VIPS-WARNING **: 06:12:34.079: Unknown field with tag 514 (0x202) encountered

(vips:84): VIPS-WARNING **: 06:12:34.079: Photometric tag is missing, assuming data is YCbCr

(vips:84): VIPS-WARNING **: 06:12:34.079: SamplesPerPixel tag is missing, applying correct SamplesPerPixel value of 3

(vips:84): VIPS-WARNING **: 06:12:34.079: Subsampling tag is not set, yet subsampling inside JPEG data [2,1] does not match default values [2,2]; assuming subsampling inside JPEG data is correct
MissingRequired: TIFF directory is missing required "ImageLength" field
tiff2vips: subsampled images not supported
```

I suppose imagemagick is doing something more.

**Comment by Tawmu** [comment]
> Olympus .orf is also broken in thumbnail and in full screen view I can provide sample file(s) if needed.

Seeing the same issue here. ORF files simply show as the broken icon. Am happy to provide example files if required.

**Comment by uhthomas** [comment]
> > Olympus .orf is also broken in thumbnail and in full screen view I can provide sample file(s) if needed.
> 
> 
> 
> Seeing the same issue here. ORF files simply show as the broken icon. Am happy to provide example files if required.

Please open a new issue.

**Comment by ixa2mka0** [comment]
Sample from Olympus XZ-1
[Olympus-XZ1.orf.zip](https://github.com/immich-app/immich/files/12652545/Olympus-XZ1.orf.zip)

**Comment by msrn** [comment]
> > > Olympus .orf is also broken in thumbnail and in full screen view I can provide sample file(s) if needed.
> > 
> > 
> > Seeing the same issue here. ORF files simply show as the broken icon. Am happy to provide example files if required.
> 
> Please open a new issue.

Here's an issue about this

https://github.com/immich-app/immich/issues/3995

**Pull request merged**